src/symbolize/gimli.rs

```diff
@@ -109,33 +151,46 @@ pub fn resolve(addr: *mut c_void, cb: &mut FnMut(&super::Symbol)) {
 
     // Finally, get a cached mapping or create a new mapping for this file, and
     // evaluate the DWARF info to find the file/line/name for this address.
-    with_mapping_for_path(path, |mapping| {
-        let (file, line, func) = match mapping.locate(addr.0 as u64) {
-            Ok(None) | Err(_) => return,
-            Ok(Some((file, line, func))) => (file, line, func),
-        };
-
-        let sym = super::Symbol {
-            inner: Symbol::new(addr.0 as usize,
-                               file,
-                               line,
-                               func.map(|f| f.to_string()))
-        };
-
-        cb(&sym);
+    with_mapping_for_path(path, |&(ref dwarf, ref symbols)| {
+        let mut found_sym = false;
+        if let Ok(mut frames) = dwarf.find_frames(addr.0 as u64) {
+            while let Ok(Some(frame)) = frames.next() {
+                let (file, line) = frame
+                    .location
+                    .map(|l| (l.file, l.line))
+                    .unwrap_or((None, None));
+                let name = frame
+                    .function
+                    .and_then(|f| f.raw_name().ok().map(|f| f.to_string()));
+                let sym = super::Symbol {
+                    inner: Symbol::new(addr.0 as usize, file, line, name),
+                };
+                cb(&sym);
+                found_sym = true;
+            }
+        }
+
+        if !found_sym {
+            if let Some(name) = symbols.get(addr.0 as u64).and_then(|x| x.name()) {
+                let sym = super::Symbol {
+                    inner: Symbol::new(addr.0 as usize, None, None, Some(name.to_string())),
+                };
+                cb(&sym);
+            }
+        }
     });
 }
 
 pub struct Symbol {
     addr: usize,
-    file: PathBuf,
+    file: Option<PathBuf>,
```

Comment:
This is optional now because the fallback might not have this info, right?

Reply:
Both that and because addr2line returns an option too, for the unlikely case where we found a function name, but not location info.